File: 1 Algorithmic Toolbox/Week 3 Greedy Algorithms/week3_greedy_algorithms/2_maximum_value_of_the_loot/fractional_knapsack.py
# Uses python3
import sys

def get_optimal_value(capacity, weights, values):
    value = 0.
    # write your code here
    Items = []
    for i in range(len(values)):
        Items.append((values[i] / weights[i], values[i], weights[i]))
        
    Items.sort(key=lambda x:x[0])
    while (capacity > 0 and len(Items) != 0):
        item = Items.pop()
        weight = min(item[2], capacity)
        value += weight * item[0]
        capacity -= weight
        
    return value


if __name__ == "__main__":
    data = list(map(int, sys.stdin.read().split()))
    n, capacity = data[0:2]
    values = data[2:(2 * n + 2):2]
    weights = data[3:(2 * n + 2):2]
    opt_value = get_optimal_value(capacity, weights, values)
    print("{:.10f}".format(opt_value))

File: 1 Algorithmic Toolbox/Week 3 Greedy Algorithms/week3_greedy_algorithms/2_maximum_value_of_the_loot/fractional_knapsack.cpp
#include <iostream>
#include <vector>
#include <algorithm>

using std::vector;

struct Item{
    int value;
    int weight;
    double ratio;
};

double get_optimal_value(int capacity, vector<int> weights, vector<int> values) {
  double value = 0.0;
  // write your code here
  std::vector<Item> Items;
  Item tmp;
  for (int i=0; i<values.size(); i++){
      tmp.value = values[i];
      tmp.weight = weights[i];
      tmp.ratio = (double)(values[i]) / (double)(weights[i]);
      Items.push_back(tmp);
  }
  
  struct compareItem{
      bool operator()(Item x, Item y){return (x.ratio <= y.ratio);}
  }mycompare;
  std::sort(Items.begin(), Items.end(), mycompare);
  
  while (capacity > 0 && Items.size() != 0){
      Item tmpitem = Items.back();
      Items.pop_back();
      int tmpweight = std::min(tmpitem.weight, capacity);
      value += (double)(tmpweight) * tmpitem.ratio;
      capacity -= tmpweight;
  }
  
  return value;
}

int main() {
  int n;
  int capacity;
  std::cin >> n >> capacity;
  vector<int> values(n);
  vector<int> weights(n);
  for (int i = 0; i < n; i++) {
    std::cin >> values[i] >> weights[i];
  }

  double optimal_value = get_optimal_value(capacity, weights, values);

  std::cout.precision(10);
  std::cout << optimal_value << std::endl;
  return 0;
}
